### ISSUE TITLE
Change C-arm to move source and detector simultaneously

### ISSUE THREAD

**Issue by eigenvivek** [eigenvivek/DiffDRR#133]
**Problem with the existing geometry implementation:** The source and detector currently move independent of each other (as is done for a standard X-ray), but in a C-arm, the source and detector have to move together. [This example](https://vivekg.dev/DiffDRR/tutorials/spherical.html)  shows that the detector moves independent of the source:

![](https://vivekg.dev/DiffDRR/tutorials/spherical_files/figure-html/cell-4-output-1.png)

Need to reimplement such that the rotation matrices are applied both to the source and detector. Also, use this opportunity to make the geometry match `DeepDRR`:

<img width="563" alt="Screen Shot 2023-04-15 at 08 31 34" src="https://user-images.githubusercontent.com/29757116/232223732-95d34da8-0d1e-43f4-be11-47e785ed28d0.png">

Note that `DeepDRR` reverses "the orbital movement and angulation" from the convention above (from [the documetation](https://github.com/arcadelab/deepdrr/blob/846e701884d44b397acdee80a71f34370ddd9db0/deepdrr/device/mobile_carm.py#L81-L110)):
```
Args:
    world_from_device: (Optional[geo.FrameTransform], optional): Transform that defines the device coordinate space in world coordinates. None is the identity transform. Defaults to None.
    isocenter (geo.Point3D): the initial isocenter of in the device frame. This is the point about which rotations are performed.
    isocenter_distance (float): the distance from the X-ray source to the isocenter of the CAarm. (The center of rotation).
    alpha (float): initial LAO/RAO angle of the C-Arm. alpha > 0 is in the RAO direction. This is the angle along arm of the Carm.
    beta (float): initial CRA/CAU angulation of the C-arm. beta > 0 is in the CAU direction.
    degrees (bool, optional): Whether given angles are in degrees. Defaults to False.
    camera_intrinsics: (Optional[Union[geo.CameraIntrinsicTransform, dict]], optional): either a CameraIntrinsicTransform instance or kwargs for CameraIntrinsicTransform.from_sizes
```

**Comment by hemmer** [comment]
It would be nice to standardise between the two libraries. I've done a bit of wrangling to map between the two (I'm sure you have a much better understanding though!), but it case it's useful:

* this issue was my attempt to reconcile the SDR variables - https://github.com/arcadelab/deepdrr/issues/80 
* if I change the order in which the rotation matrices are applied in DiffDRR AND use the mapping below, I get consistent behaviour (or put another can reproduce the pose of one library from the other):


**Mapping to DeepDRR conventions**
```
carm = MobileCArm(isocenter=volume.center_in_world, alpha=theta, beta=np.pi / 2 - phi, gamma=-gamma
```

**Change Matrix Application Order in DiffDRR**

```
diff --git a/DiffDRR/diffdrr/camera.py b/DiffDRR/diffdrr/camera.py
index dadcce9..6604b99 100644
--- a/DiffDRR/diffdrr/camera.py
+++ b/DiffDRR/diffdrr/camera.py
@@ -92,7 +92,7 @@ def Rxyz(rotations):
     R_z = Rz(theta, batch_size, device)
     R_y = Ry(phi, batch_size, device)
     R_x = Rx(gamma, batch_size, device)
-    return torch.einsum("bij,bjk,bkl->bil", R_z, R_y, R_x)
+    return torch.einsum("bij,bjk,bkl->bil", R_y, R_z, R_x)
 
 
 def Rx(gamma, batch_size, device):
```

If I make those changes / mappings, I can get consistent results (up to expected minor differences due to the way each renderer works):

![image](https://user-images.githubusercontent.com/915048/233951930-4273c70c-16e7-448f-806b-5873e93468c2.png)

**Comment by hemmer** [comment]
(to be clear I'm not saying what I've done is necessarily the correct way to map between them, but it's the only way I've managed to get consistent results so far).

**Comment by eigenvivek** [comment]
@hemmer Thank you very much for sharing this! I will retry your code with real CT. If all works well, I'll PR the changes in!

**Comment by eigenvivek** [comment]
@hemmer Works perfectly! thanks very much for all the heavy lifting on your end. I would never have thought to reverse the order of matrix multiplications. I'll include a notebook showing how to do the conversion from DiffDRR -> DeepDR in the documentation.

<img width="591" alt="PNG image" src="https://user-images.githubusercontent.com/29757116/234665558-f6ab5213-a180-4155-b0b2-460aa9a7fae1.png">

**Comment by fedeface98** [comment]
Hello,

I did not understand well the mapping. In deepDRR I put as alpha the primary angle and as beta the secondary angle and gamma is 0. If I want to reproduce the same results with diffDRR in the following code gamma = alpha and phi = beta ?
@@ -92,7 +92,7 @@ def Rxyz(rotations):
     R_z = Rz(theta, batch_size, device)
     R_y = Ry(phi, batch_size, device)
     R_x = Rx(gamma, batch_size, device)
-    return torch.einsum("bij,bjk,bkl->bil", R_z, R_y, R_x)
+    return torch.einsum("bij,bjk,bkl->bil", R_y, R_z, R_x)

**Comment by eigenvivek** [comment]
Hi @fedeface98 , this part of the documentation should cover how to do the conversions: https://vivekg.dev/DiffDRR/tutorials/metamorphasis.html

Please ping me if any part doesn't make sense or doesn't work.

**Comment by fedeface98** [comment]
in this part for you alpha and beta are the primary and secondary angles?
I can't understand why you need to do np.pi / 2 - beta

def test_phantom_deepdrr(alpha, beta, gamma):
    with Projector(
        volume=patient,
        carm=carm,
    ) as projector:
        carm.move_to(
            alpha=np.rad2deg(alpha),
            beta=np.rad2deg(np.pi / 2 - beta),
            gamma=np.rad2deg(-gamma),
            degrees=True,
        )
        img = (
            projector()
        )  # The first run doesn't use updated parameters, for some reason?
        img = projector()[:, ::-1].copy()
    return img

**Comment by eigenvivek** [comment]
yes alpha and beta are the primary and secondary angles, respectively

`np.pi / 2  - beta` is a correction factor used because `DiffDRR` initializes the camera on the x-axis while `DeepDRR` initializes it on the z-axis

**Comment by fedeface98** [comment]
Thank you, it is clear to me!

But then in the following function you inverse beta and alpha, why?

def diffdrr_to_deepdrr(euler_angles):
    alpha, beta, gamma = euler_angles.unbind(-1)
    return torch.stack([beta, alpha, gamma], dim=1)

**Comment by eigenvivek** [comment]
i'm sorry, i probably misspoke --- i should have said `beta` is the primary angle and `alpha` is the secondary angle

the inversions aren't special in any way. they're simply helper functions to convert from `diffdrr`'s convention to `deepdrrr`'s convention

note that when using euler angles to parameterize a rotation, there are 12 equivalent conventions you can use (see [this table](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix)). `diffdrr` uses one and `deepdrr` uses another, so the functions just perform some conversions

**Comment by fedeface98** [comment]
Thanks a lot!

I have a final question about the isocenter. In deepdrr the isocenter is put equal to patient.center_in_world that is different from the isocenter set up in diffdrr calculated as (torch.tensor(volume.shape) - 1) * torch.tensor(spacing) / 2).

I can't understand why.
